homeassistant/components/mqtt/fan.py

```diff
@@ -338,137 +334,142 @@ def _setup_from_config(self, config: ConfigType) -> None:
+        try:
+            percentage = ranged_value_to_percentage(
+                self._speed_range, int(rendered_percentage_payload)
+            )
+        except ValueError:
+            _LOGGER.warning(
+                (
+                    "'%s' received on topic %s. '%s' is not a valid speed within"
+                    " the speed range"
+                ),
+                msg.payload,
+                msg.topic,
+                rendered_percentage_payload,
+            )
+            return
+        if percentage < 0 or percentage > 100:
+            _LOGGER.warning(
+                (
+                    "'%s' received on topic %s. '%s' is not a valid speed within"
+                    " the speed range"
+                ),
+                msg.payload,
+                msg.topic,
+                rendered_percentage_payload,
+            )
+            return
+        self._attr_percentage = percentage
+
+    @callback
+    def _preset_mode_received(self, msg: ReceiveMessage) -> None:
+        """Handle new received MQTT message for preset mode."""
+        preset_mode = str(self._value_templates[ATTR_PRESET_MODE](msg.payload))
+        if preset_mode == self._payload["PRESET_MODE_RESET"]:
+            self._attr_preset_mode = None
+            return
+        if not preset_mode:
+            _LOGGER.debug("Ignoring empty preset_mode from '%s'", msg.topic)
+            return
+        if not self.preset_modes or preset_mode not in self.preset_modes:
+            _LOGGER.warning(
+                "'%s' received on topic %s. '%s' is not a valid preset mode",
+                msg.payload,
+                msg.topic,
+                preset_mode,
+            )
+            return
+
+        self._attr_preset_mode = preset_mode
+
+    @callback
+    def _oscillation_received(self, msg: ReceiveMessage) -> None:
+        """Handle new received MQTT message for the oscillation."""
+        payload = self._value_templates[ATTR_OSCILLATING](msg.payload)
+        if not payload:
+            _LOGGER.debug("Ignoring empty oscillation from '%s'", msg.topic)
+            return
+        if payload == self._payload["OSCILLATE_ON_PAYLOAD"]:
+            self._attr_oscillating = True
+        elif payload == self._payload["OSCILLATE_OFF_PAYLOAD"]:
+            self._attr_oscillating = False
+
+    @callback
+    def _direction_received(self, msg: ReceiveMessage) -> None:
+        """Handle new received MQTT message for the direction."""
+        direction = self._value_templates[ATTR_DIRECTION](msg.payload)
+        if not direction:
+            _LOGGER.debug("Ignoring empty direction from '%s'", msg.topic)
+            return
+        self._attr_current_direction = str(direction)
+
     def _prepare_subscribe_topics(self) -> None:
         """(Re)Subscribe to topics."""
         topics: dict[str, Any] = {}
 
-        def add_subscribe_topic(topic: str, msg_callback: MessageCallbackType) -> bool:
+        def add_subscribe_topic(
```

Comment:
At some point it would be nicer to refactor it to work like humidifier and not have the inner function here https://github.com/home-assistant/core/pull/118116/files#diff-8510c9670b2ac7992ebbe838901a10a063b3e7eaf1f4e5d94b78102c8cac04b7R422